File: companies/Oracle/36.valid-sudoku.java
import java.util.*;

/**
 * Algorithm:
 * Check each filled digit against the row, column, and 3x3 box sets that have
 * already seen digits. A repeated digit in any unit makes the board invalid.
 *
 * Java data structures:
 * HashSet<Character>[] stores seen digits for nine rows, columns, and boxes.
 *
 * Complexity:
 * O(81) time and O(81) space, both constant for a 9x9 board.
 */
class Solution {
    public boolean isValidSudoku(char[][] board) {
        Set<Character>[] rows = new HashSet[9];
        Set<Character>[] cols = new HashSet[9];
        Set<Character>[] boxes = new HashSet[9];
        for (int i = 0; i < 9; i++) {
            rows[i] = new HashSet<>();
            cols[i] = new HashSet<>();
            boxes[i] = new HashSet<>();
        }

        for (int r = 0; r < 9; r++) {
            for (int c = 0; c < 9; c++) {
                char val = board[r][c];
                if (val == '.') {
                    continue;
                }
                int b = (r / 3) * 3 + c / 3;
                if (rows[r].contains(val) || cols[c].contains(val) || boxes[b].contains(val)) {
                    return false;
                }
                rows[r].add(val);
                cols[c].add(val);
                boxes[b].add(val);
            }
        }
        return true;
    }
}

File: companies/WEX/36.valid-sudoku.go
package leetcode

// IsValidSudoku36 tracks seen digits for each row, column, and 3x3 box. In Go,
// fixed [9][9]bool arrays are more direct than hash sets because the board and
// digit domain are constant size.
//
// Time/space: O(1) for the fixed 9x9 board.
func IsValidSudoku36(board [][]byte) bool {
	var rows, cols, boxes [9][9]bool
	for r := 0; r < 9; r++ {
		for c := 0; c < 9; c++ {
			val := board[r][c]
			if val == '.' {
				continue
			}
			d := val - '1'
			b := (r/3)*3 + c/3
			if rows[r][d] || cols[c][d] || boxes[b][d] {
				return false
			}
			rows[r][d], cols[c][d], boxes[b][d] = true, true, true
		}
	}
	return true
}
